### DISCUSSION TITLE
Project Structure & Repo Guidelines Discussion

### DISCUSSION THREAD

**Discussion by Yonodactyl** [Godot-Community-Games-Org/game1#4] [General]
As we’re gearing up to start building our game in Godot, I wanted to kick off a quick discussion on how we want to structure things in the repo and set some ground rules to keep everything running smoothly.

## Things to Consider:

**How should we organize the Godot project files?**
- 🗂 Do we want separate folders for scenes, assets, scripts, etc.?
- Should we follow any specific naming conventions?

**How do we handle new features or bug fixes?**
- Create a branch for each feature?
- Or, do we just PR directly into main?
- Any specific workflow we want to stick to?

**Do we want to keep commit messages loose, or follow a more standardized style (e.g., “fix:”, “feat:”)?**

**Are there any Godot plugins or dependencies we know we’ll need off the bat?**

> Let’s make sure we have a process for adding/removing things so the repo doesn’t blow up.

**How strict should we be with code reviews?**
- Is it “submit whenever” or do we want at least one approval before merging?

**How should we handle the artwork, sounds, etc.?**
- Any best practices for keeping file sizes down, or should we just sync the repo and see what happens?

Feel free to throw in any other ideas or concerns you have. Let’s figure this stuff out early so we can focus on the fun part—making the game!

**Comment by ryankmims** [comment]
> Do we want separate folders for scenes, assets, scripts, etc.?

Yeah I think following the typical Unity style would be good. So Scenes are "levels" basically. Assets would contain things like materials/shaders. I like the idea of a Components folder which contains any packed scene that goes into a world that functions independently. So like: 'Components/Player/Player.tscn' or 'Components/Mobs/Zombie.tscn' etc. -- and then Scripts are in their own folder, structured the same way: 'Scripts/Player/Player.gd' or 'Scripts/Mobs/Zombie.gd'

We don't have to do it that way. But I think it's personally easier to find stuff.

> Create a branch for each feature?

Personally, I vote 100% yes on this. I like the branch structure of doing main -> development -> release and then the main branch contains the most stable release once it's been tested. Then on the development branch you have feature branches that work toward the goals we're trying to reach for that development cycle. 

That's how I've done it as an engineer. But each development cycle was scheduled around fiscal quarters. I don't know how we'd do that here with an open source project

> Do we want to keep commit messages loose, or follow a more standardized style (e.g., “fix:”, “feat:”)?

Probably short/succinct like you have listed

> How strict should we be with code reviews?

Yeah that's going to be the toughest part of this I think. I want to say really strict. But at the very least you should need one approval. If you're pushing a fix to something, maybe not as strict (like a quick glance at the code). If you're adding a feature then I feel like it needs to be pulled by someone and tested.

But that could turn into a headache really quickly, so I don't know.

**Reply by ryankmims** [reply]
Oh, and a random concern I just had is what would our policy be for using the "tool" annotation? 

Like I doubt we want every single script to be a tool that's running something every frame in the editor.

**Reply by LightInn** [reply]
for the branche/commit part, I would say : 

we can use tags on Main to mark release, otherwise people have their "feat" branch, making PR to main 

On theirs own branch, people can name commit as they want, but when they merge a pr, it could be a rebase&merge+squashing with a proper naming, so the main branch is always clean

**Reply by BlodyxCZ** [reply]
I totally agree on the branches, but I wonder how many features branches there will be as more people join in

**Reply by Yonodactyl** [reply]
> On theirs own branch, people can name commit as they want, but when they merge a pr, it could be a rebase&merge+squashing with a proper naming, so the main branch is always clean

If we are going the route of rebasing and squashing, it might be worth having a quick blurb somewhere to inform new contributors and also help those not familiar with the process. I like this idea, I just can imagine the nightmare of a rebase if they are not familiar with how it works!

**Reply by LightInn** [reply]
I dont know if there is some kind of "tests" we could run in CI to help validate PR

**Reply by Yonodactyl** [reply]
> I dont know if there is some kind of "tests" we could run in CI to help validate PR

I'm very certain that Godot has testing frameworks and that we can have verified with CI. Never done that before on my projects, but would love to see it done!

**Comment by GodotCommunityGames** [comment]
Yes these are all crucial questions to figure out and lay out answers for before we start working on games.

And once these decisions get made I'll add all decisions to the organization README so that all the guidelines are clear.


Having a standard file structure I think would be a very good idea. At the start of each game have all the folders of the file structure set up so doesn't need to be guessing where files should go. I'm just a month into Godot so I'm still hashing out my file structure preferences, but I've been doing something along the lines of:

/assets
--/art
----/tilesets
----/spritesheets
----/sprites
--/music
--/audio
/scenes
--/characters
----player
----/enemies
----/bosses
----/mini_bosses (if a game has them)
----/npcs
--/levels
--/items
--/weapons
--/effects
--/components (for composable components)
/scripts (for standalone scripts and classes that aren't paired with a scene)
--/autoloads

That's just the structure I've started developing. But yeah we'll have to figure out what the community wants to settle on.


Naming conventions just follow the Godot naming conventions. For any C# stuff I think there's slightly different naming conventions like for file names I believe which have to be the same as the class name in the file, but in general I think the naming conventions in godot are: variable names and signal names and function snake_case, constants UPPERCASE, class names PascalCase, file names and folder names are snake_case I think.

Yeah I think we should definitely follow standard branch procedure of people working on features in branches and merging when done. PR directly into main would probably be a nightmare. And we'll have to set up issues for each repo where people take on specific features or parts of features and bug fixes and whatnot.

And good question on commit messages. I've always done messages informally but could be good to have some formal guidelines to easily label commit types.

Code review is gonna be an interesting issue, agreed with ryan that may be the toughest part logistically. Does each project have a formal review and merge process that perhaps more experienced devs or project leaders manage, or more informal of people just merging when things are ready. Also this could be a decision to be made on a game by game basis. There may be big games that a bunch of people work on that requires a more formal review process, or just there may be someone who want to take the project lead on a game and perform that duty, and then there may be smaller games that just a few people work on where they don't even bother with code reviews.

Figuring out some agreed upon best practices would be good too, yeah. Both for keeping games consistent and for plenty of us to learn what some best practices are!

**Comment by Yonodactyl** [comment]
I'm not familiar with scripting and code generation, but it would be **_magical_** if we had some sort of script we can run to generate the template of the games set up using the standards we decide on.

For example:
Oh you're building a roguelike 2D pixel art game? Run this script and Godot will set up the project to use pixel snapping, canvas viewport, a default size, and change filtering for all imports to nearest. It will also set up the folders in the project.

That would be MAGICAL! I can take this on to start researching!

**Reply by GodotCommunityGames** [reply]
yeah that'd be a cool thing. Back in my web dev days I totally had scripts to set up my whole directory structure and basic templates to start out with. Let us know how your research goes in figuring out that stuff.

**Reply by KaizNike** [reply]
Github provides Github templates to make template projects anyone can copy to make games out of. Could use that.
